src/pybind/mgr/cli_api_commands/module.py

```diff
@@ -0,0 +1,233 @@
+        self.signature = inspect.signature(func)
+        self.params = self.signature.parameters
+
+    def to_ceph_signature(self):
+        '''
+        Generate CephCommand signature (dict-like)
+        '''
+        return {
+            'prefix': f'mgr cli {self.func.__name__}',
+            'perm': API.perm.get(self.func)
+        }
+
+    def to_ceph_benchmark(self):
+        '''
+        Generate CephCommand signature (dict-like)
+        '''
+        return {
+            'prefix': f'mgr cli benchmark {self.func.__name__}',
+            'perm': API.perm.get(self.func)
+        }
+
+
+class ThreadedBenchmarkRunner:
+    def __init__(self, func_name: str, data: list, number_of_total_calls: int, number_of_parallel_calls: int):
+        self._func_name = func_name
+        self._data = data
+        self._number_of_total_calls = number_of_total_calls
+        self._number_of_parallel_calls = number_of_parallel_calls
+        self._threads = []
+        self._jobs: Queue = Queue()
+        self.time_accumlated = 0.0
+        self._time = []
+        self._lock = threading.Lock()
+        self._mgr = None
+
+    def set_mgr(self, mgr):
+        self._mgr = mgr
+
+    def benchmark(self):
+        self.start(self._func_name, self._data)
+        self.join()
+        stats = self.get_stats()
+        return HandleCommandResult(0, json.dumps(stats), "Error")
+
+    def start(self, func, args):
+        self.args = args
+        if(self._number_of_total_calls and self._number_of_parallel_calls):
+            for thread_id in range(self._number_of_parallel_calls):
+                new_thread = threading.Thread(target=self.timer,
+                                              args=(self._jobs, func,))
+                self._threads.append(new_thread)
+            for job_id in range(self._number_of_total_calls):
+                self._jobs.put(job_id)
+            for thread in self._threads:
+                thread.start()
+        else:
+            raise BenchmarkException("Number of Total and number of parallel calls must be greater than 0")
+
+    def join(self):
+        for thread in self._threads:
+            thread.join()
+
+    def get_stats(self):
+        stats = {
+            "avg": (self.time_accumlated / self._number_of_total_calls),
+            "min": min(self._time),
+            "max": max(self._time)
+        }
+        return stats
+
+    def timer(self, jobs, func):
+        self._lock.acquire()
+        while not self._jobs.empty():
+            jobs.get()
+            time_start = time.time()
+            if len(self.args) == 0:
+                func(self._mgr)
+            elif len(self.args) == 1:
+                func(self._mgr, self.args[0])
+            else:
+                func(self._mgr, self.args)
+            time_end = time.time()
+            self._time.append(time_end - time_start)
+            self.time_accumlated += (time_end - time_start)
+            self._jobs.task_done()
+        self._lock.release()
```

Comment:
@Waadkh7 I've been checking this function and it's "over-locking": queues and list.append are thread-safe, so you don't need to lock anything here:
```py
    def timer(self, func):
        while not self._jobs.empty():
            self._jobs.get()
            time_start = get_time()
            if len(self.args) == 0:
                func(self._mgr)
            elif len(self.args) == 1:
                func(self._mgr, self.args[0])
            else:
                func(self._mgr, self.args)
            self._time.append(get_time() - time_start)
            self._jobs.task_done()
```

The interesting side-effect of this is that, without locking, the results are worse:
```sh
# BEFORE (with all-locked timer)
# time ceph mgr cli benchmark get osd_map 10000 8
{
    "avg": 0.0006781596473505487,
    "max": 0.015568311995593831,
    "min": 0.0005710289988201112
}

real	0m7.606s
user	0m0.623s
sys	0m0.036s

```
```sh
# AFTER (only self._time locked)
# time ceph mgr cli benchmark get osd_map 10000 8
{
    "avg": 0.004328038047239534,
    "max": 1.5950815029791556,
    "min": 0.0005581160075962543
}

real	0m7.319s
user	0m0.587s
sys	0m0.045s

```

With the all-locked version, the average and worse cases are better (0.7 ms vs. 4.3 ms, and 15 ms vs. 1595 ms).

This can be explained because the overhead of contention in the C++ was a real issue, as we suspected from the field reports. We couldn't properly reproduce this when we synthetically tested the TTL cache, since this locking was indirectly alleviating the problem.

This last result is consistent with a different much simpler approach (`concurrent.futures.ThreadExecutor`):
```py
    @CLICommand('mgr cli_benchmark')
    def benchmark(self, func_name: str, func_args: List[str], _end_positional_: int = 0,
                  iterations: Optional[int]=1, threads: Optional[int]=1) -> HandleCommandResult:
        try:
            func = getattr(self, func_name)
        except:
            return HandleCommandResult(stderr="Could not find the public function you are requesting")

        def timer(i):
            time_start = get_time()
            func(*func_args)
            return get_time() - time_start

        with concurrent.futures.ThreadPoolExecutor(max_workers=threads) as executor:
            results_iter = executor.map(timer, range(iterations))
        results = list(results_iter)

        stats = {
            "avg": sum(results)/len(results),
            "max": max(results),
            "min": min(results),
        }
        return HandleCommandResult(stdout=json.dumps(stats, indent=4))
```
```sh
{
    "avg": 0.003555743730676477,
    "max": 0.9242840899969451,
    "min": 0.0006099830206949264,
    "count": 10000
}

real	0m9.329s
user	0m0.616s
sys	0m0.037s
```